engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java

```diff
@@ -1738,38 +1738,58 @@ public Class<?> visit(MethodCallExpr n, VisitArgs printer) {
-                if (params.size() != expressions.length) {
-                    throw new RuntimeException("Numba vectorized function argument count mismatch: " + params.size()
-                            + " vs." + expressions.length);
-                }
-                for (int i = 0; i < expressions.length; i++) {
-                    if (!(expressions[i] instanceof NameExpr)) {
-                        throw new RuntimeException("Numba vectorized function arguments can only be columns.");
-                    }
-                    if (!isSafelyCoerceable(expressionTypes[i], params.get(i))) {
-                        throw new RuntimeException("Numba vectorized function argument type mismatch: "
-                                + expressionTypes[i].getSimpleName() + " -> " + params.get(i).getSimpleName());
-                    }
+                CallableWrapper callableWrapper = (CallableWrapper) param.getValue();
+                try {
+                    checkVectorizability(n, expressions, expressionTypes, callableWrapper);
+                } catch (RuntimeException ex) {
+                    if (callableWrapper.isVectorized())
+                        throw ex;
                 }
+                break;
+            }
+        }
+    }
+
+    private void checkVectorizability(MethodCallExpr n, Expression[] expressions, Class<?>[] expressionTypes,
+            CallableWrapper callableWrapper) {
+
+        callableWrapper.parseSignature();
+
+        n.getParentNode().ifPresent(parent -> {
+            if (parent.getClass() != CastExpr.class || parent.getParentNode().isPresent()) {
+                throw new RuntimeException("Python vectorized function can't be used in an expression.");
+            }
+            // TODO what to do with 'cast'?
+        });
+
+        List<Class<?>> params = callableWrapper.getParamTypes();
+        if (params.size() != expressions.length) {
+            throw new RuntimeException("Python vectorized function argument count mismatch: " + params.size()
+                    + " vs." + expressions.length);
+        }
+        for (int i = 0; i < expressions.length; i++) {
+            if (!(expressions[i] instanceof NameExpr)) {
+                throw new RuntimeException("Python vectorized function arguments can only be columns.");
+            }
+            if (!isSafelyCoerceable(expressionTypes[i], params.get(i))) {
+                throw new RuntimeException("Python vectorized function argument type mismatch: "
+                        + expressionTypes[i].getSimpleName() + " -> " + params.get(i).getSimpleName());
             }
         }
+
+        callableWrapper.setVectorizable(true);
     }
 
+
     private static boolean isSafelyCoerceable(Class<?> expressionType, Class<?> aClass) {
         // TODO (core#709): numba does appear to check for type coercing at runtime, though no explicit rules exist
         return true;
```

Comment:
Type coercing check and evaluation time casting should not only benefit vectorization. But it is something that's really good to have but not a must-have because the users can always make a conscious effort to guarantee the types match, and it is not a small effort. I think for now we can keep the strategy as is, i.e. let the mismatch be caught at runtime or show up in the result.  I added a bullet point to address this in the future.

Reply:
Type coercing check and evaluation time casting should not only benefit vectorization. But it is something that's really good to have but not a must-have because the users can always make a conscious effort to guarantee the types match, and it is not a small effort. I think for now we can keep the strategy as is, i.e. let the mismatch be caught at runtime or show up in the result.  I added a bullet point to address this in the future.